alf/trainers/off_policy_trainer.py

```diff
@@ -0,0 +1,166 @@
+            total steps from all individual environment in the batch
+            environment.
+        use_tf_functions (bool): whether to use tf.function
+        summary_interval (int): write summary every so many training steps (
+            i.e. number of parameter updates)
+        summaries_flush_secs (int): flush summary to disk every so many seconds.
+        eval_interval (int): evaluate every so many iteration
+        num_eval_episodes (int) : number of episodes for one evaluation
+        checkpoint_interval (int): checkpoint every so many iterations
+        debug_summaries (bool): A bool to gather debug summaries.
+        summarize_grads_and_vars (bool): If True, gradient and network variable
+            summaries will be written during training.
+    """
+
+    train_dir = os.path.expanduser(train_dir)
+    eval_dir = os.path.join(os.path.dirname(train_dir), 'eval')
+    num_steps_per_iter = (
+        math.ceil(num_steps_per_iter / (mini_batch_length * env.batch_size)) *
+        mini_batch_length * env.batch_size)
+    eval_metrics = None
+    eval_summary_writer = None
+    if eval_env is not None:
+        eval_metrics = [
+            tf_metrics.AverageReturnMetric(buffer_size=num_eval_episodes),
+            tf_metrics.AverageEpisodeLengthMetric(
+                buffer_size=num_eval_episodes)
+        ]
+        eval_summary_writer = tf.summary.create_file_writer(
+            eval_dir, flush_millis=summaries_flush_secs * 1000)
+
+    def train_():
+        tf.random.set_seed(random_seed)
+        global_step = get_global_counter()
+
+        driver = OffPolicyDriver(
+            env=env,
+            algorithm=algorithm,
+            debug_summaries=debug_summaries,
+            summarize_grads_and_vars=summarize_grads_and_vars)
+        replay_buffer = driver.add_replay_buffer()
+
+        checkpointer = tfa_common.Checkpointer(
+            ckpt_dir=os.path.join(train_dir, 'algorithm'),
+            algorithm=algorithm,
+            metrics=metric_utils.MetricsGroup(driver.get_metrics(), 'metrics'),
+            global_step=global_step)
+        checkpointer.initialize_or_restore()
+
+        if use_tf_functions:
+            driver.run = tf.function(driver.run)
+            driver.train = tf.function(driver.train)
+
+        env.reset()
+        time_step = driver.get_initial_time_step()
+        policy_state = driver.get_initial_state()
+        for iter in range(num_iterations):
+            t0 = time.time()
+            time_step, policy_state = driver.run(
+                max_num_steps=num_steps_per_iter,
+                time_step=time_step,
+                policy_state=policy_state)
+
+            experience = replay_buffer.gather_all()
+            driver.train(
+                experience,
+                num_updates=num_updates_per_train_step,
+                mini_batch_length=mini_batch_length,
+                mini_batch_size=mini_batch_size)
+            replay_buffer.clear()
```

Comment:
it performs on policy training if replay buffer was cleared every iteration
why can not sample a batch from the replay buffer and perform an update as `off_policy_driver_test` do ?